File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.js
const fs = require("fs");
const input = fs.readFileSync('/dev/stdin').toString().split("\n");

const N = Number(input[0]);
const matrix = []
for (let i = 1; i <= N; i++) {
  let line = input[i].split(" ").map(Number);
  matrix.push(line);
}

// 3차원 배열: dp[y][x][direction]
// direction: 0=가로, 1=세로, 2=대각선
let dp = Array.from({ length: N }, () => 
  Array.from({ length: N }, () => new Array(3).fill(-1))
);

function dfs(x, y, direction) {
  if (x == N - 1 && y == N - 1) return 1;

  // 이미 계산한 상태면 반환
  if (dp[y][x][direction] !== -1) {
    return dp[y][x][direction];
  }

  let count = 0;

  if (direction === 0) { // 가로
    if (x + 1 < N && matrix[y][x + 1] == 0) {
      count += dfs(x + 1, y, 0);
    }
    if (x + 1 < N && y + 1 < N && 
        matrix[y][x + 1] == 0 && matrix[y + 1][x] == 0 && matrix[y + 1][x + 1] == 0) {
      count += dfs(x + 1, y + 1, 2);
    }
  } else if (direction === 1) { // 세로
    if (y + 1 < N && matrix[y + 1][x] == 0) {
      count += dfs(x, y + 1, 1);
    }
    if (x + 1 < N && y + 1 < N && 
        matrix[y][x + 1] == 0 && matrix[y + 1][x] == 0 && matrix[y + 1][x + 1] == 0) {
      count += dfs(x + 1, y + 1, 2);
    }
  } else if (direction === 2) { // 대각선
    if (x + 1 < N && matrix[y][x + 1] == 0) {
      count += dfs(x + 1, y, 0);
    }
    if (y + 1 < N && matrix[y + 1][x] == 0) {
      count += dfs(x, y + 1, 1);
    }
    if (x + 1 < N && y + 1 < N && 
        matrix[y][x + 1] == 0 && matrix[y + 1][x] == 0 && matrix[y + 1][x + 1] == 0) {
      count += dfs(x + 1, y + 1, 2);
    }
  }

  dp[y][x][direction] = count;
  return count;
}

console.log(dfs(1, 0, 0));

File: 백준/Gold/17070. 파이프 옮기기 1/파이프 옮기기 1.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

n = int(input())
matrix = [list(map(int, input().split())) for _ in range(n)]
memo = {}

def dfs(x, y, direction):
    # 목표 지점 도착
    if x == n - 1 and y == n - 1:
        return 1
    
    # 메모이제이션
    if (x, y, direction) in memo:
        return memo[(x, y, direction)]
    
    count = 0
    
    # 가로 방향 (direction == 0)
    if direction == 0:
        # 가로 -> 가로
        if x + 1 < n and matrix[y][x + 1] == 0:
            count += dfs(x + 1, y, 0)
        # 가로 -> 대각선
        if (x + 1 < n and y + 1 < n and 
            matrix[y][x + 1] == 0 and matrix[y + 1][x] == 0 and matrix[y + 1][x + 1] == 0):
            count += dfs(x + 1, y + 1, 2)
    
    # 세로 방향 (direction == 1)
    elif direction == 1:
        # 세로 -> 세로
        if y + 1 < n and matrix[y + 1][x] == 0:
            count += dfs(x, y + 1, 1)
        # 세로 -> 대각선
        if (x + 1 < n and y + 1 < n and 
            matrix[y][x + 1] == 0 and matrix[y + 1][x] == 0 and matrix[y + 1][x + 1] == 0):
            count += dfs(x + 1, y + 1, 2)
    
    # 대각선 방향 (direction == 2)
    else:
        # 대각선 -> 가로
        if x + 1 < n and matrix[y][x + 1] == 0:
            count += dfs(x + 1, y, 0)
        # 대각선 -> 세로
        if y + 1 < n and matrix[y + 1][x] == 0:
            count += dfs(x, y + 1, 1)
        # 대각선 -> 대각선
        if (x + 1 < n and y + 1 < n and 
            matrix[y][x + 1] == 0 and matrix[y + 1][x] == 0 and matrix[y + 1][x + 1] == 0):
            count += dfs(x + 1, y + 1, 2)
    
    memo[(x, y, direction)] = count
    return count

print(dfs(1, 0, 0))  # 초기: (1, 0) 위치, 가로 방향